wpilibj/src/main/java/edu/wpi/first/wpilibj/RobotBase.java

```diff
@@ -221,44 +218,24 @@ public static void initializeHardwareConfiguration() {
    */
   @SuppressWarnings({"PMD.AvoidInstantiatingObjectsInLoops", "PMD.AvoidCatchingThrowable",
                      "PMD.CyclomaticComplexity", "PMD.NPathComplexity"})
-  public static void main(String... args) {
+  public static <T extends RobotBase> void startRobot(Class<T> cls) {
```

Comment:
This doesn't need to be a parameterized method, just do
```java
public static void startRobot(Class<? extends RobotBase> cls)
```

There also should be a check to see if it's a valid class (ie not `RobotBase` or any of the built-in subclasses). Just checking for the `abstract` modifier flag should be enough. `cls.newInstance()` will throw an exception if `cls` is abstract, but it'd be better if we checked and throw an error with a message. The default error has no message, so it's not at all clear what the user did wrong.

Reply:
@SamCarlberg The generic is needed to get T on line 230.

Reply:
As for the abstract thing I can do that. Theres no way to make T not abstract in Java, right?

Reply:
You don't need `T` at all, though

Reply:
Just declare it as `RobotBase`. That's what it gets compiled to anyway

Reply:
Instead of having a parameter of a class, you could have a parameter of a supplier to make sure you get an object.

Reply:
But then the user has to provide that supplier, so No.

Reply:
Does removing the generic save anything? I like the explicit T better in a lot of ways. It looks a lot better.

Reply:
And we don't want to provide an object to this. We need initialization to happen before construction.

Reply:
```
public static void main(String[] args) {
  startRobot(MyRobot::new);
}
```
```
public static <T extends RobotBase> void startRobot(Supplier<T> robotSupplier) {
...
  T robot;
  try {
  robot = robotSupplier.get();
...
}
```

Reply:
Does that initialize the robot at the get()? If so I do like that better because it requires a parameterless main.

Reply:
This also allows the user to pass constructor parameters if they wanted to do that.

Reply:
Again, you don't need to parameterize the method. It's only useful if you have multiple generic arguments that need to share the same type eg `<T> void copy(List<T> src, List<T> dst)` or if you're returning a value dependent on the arguments eg `<T> T get(Supplier<T> supplier)`.

Since neither case applies, the generic parameters can be removed and you can use a wildcard bounded generic:

```java
public static void startRobot(Supplier<? extends RobotBase> robotSupplier) {
   ...
}
```

Reply:
I still dont agree, because maybe someday later we actually want T, and then its a breaking change to the byte code. Plus the parameterized method looks better and is easier to understand for developers.